mne/inverse_sparse/mxne_optim.py

```diff
@@ -322,78 +387,206 @@ def mixed_norm_solver(M, G, alpha, maxit=3000, tol=1e-8, verbose=None,
-
-        active_set = np.zeros_like(active_set)
-        active_set[as_] = True
     else:
-        X, active_set, E = l21_solver(M, G, alpha, maxit=maxit,
-                                      tol=tol, n_orient=n_orient)
+        X, active_set, E = l21_solver(M, G, alpha, lc, maxit=maxit,
+                                      tol=tol, n_orient=n_orient, init=None)
+
+    if (active_set.sum() > 0) and debias:
+        bias = compute_bias(M, G[:, active_set], X, n_orient=n_orient)
+        X *= bias[:, np.newaxis]
+
+    logger.info('Final active set size: %s' % (np.sum(active_set) // n_orient))
+
+    return X, active_set, E
+
+
+@verbose
+def iterative_mixed_norm_solver(M, G, alpha, n_mxne_iter, maxit=3000,
+                                tol=1e-8, verbose=None, active_set_size=50,
+                                debias=True, n_orient=1, solver='auto'):
+    """Solves L0.5/L2 mixed-norm inverse problem with active set strategy
+
+    Algorithm is detailed in:
+    Strohmeier D., Haueisen J., and Gramfort A.:
+    Improved MEG/EEG source localization with reweighted mixed-norms,
+    4th International Workshop on Pattern Recognition in Neuroimaging,
+    Tuebingen, 2014
+
+    Parameters
+    ----------
+    M : array, shape (n_sensors, n_times)
+        The data.
+    G : array, shape (n_sensors, n_dipoles)
+        The gain matrix a.k.a. lead field.
+    alpha : float
+        The regularization parameter. It should be between 0 and 100.
+        A value of 100 will lead to an empty active set (no active source).
+    n_mxne_iter : int
+        The number of MxNE iterations. If > 1, iterative reweighting
+        is applied.
+    maxit : int
+        The number of iterations.
+    tol : float
+        Tolerance on dual gap for convergence checking.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+    active_set_size : int
+        Size of active set increase at each iteration.
+    debias : bool
+        Debias source estimates.
+    n_orient : int
+        The number of orientation (1 : fixed or 3 : free or loose).
+    solver : 'prox' | 'cd' | 'bcd' | 'auto'
+        The algorithm to use for the optimization.
+
+    Returns
+    -------
+    X : array, shape (n_active, n_times)
+        The source estimates.
+    active_set : array
+        The mask of active sources.
+    E : list
+        The value of the objective function over the iterations.
+    """
+    g = lambda w: np.sqrt(np.sqrt(groups_norm2(w.copy(), n_orient)))
+    gprime = lambda w: 2. * np.repeat(g(w), n_orient).ravel()
+
+    E = list()
+
+    active_set = np.ones(G.shape[1], dtype=np.bool)
+    weights = np.ones(G.shape[1])
+    X = np.zeros((G.shape[1], M.shape[1]))
+
+    for k in range(n_mxne_iter):
+        X0 = X.copy()
+        active_set_0 = active_set.copy()
+        G_tmp = G[:, active_set] * weights[np.newaxis, :]
+
+        if k == 0:
```

Comment:
We should change the if statement here. We could either delete it completely and always use the first option or we test if the number of active sources is greater than active_set_size (than we use the first option with active_set_size=active_set_size) or smaller (than we use the second without active_set_size=None)